File: src/main/java/org/dbtools/query/shared/filter/LikeFilter.java
package org.dbtools.query.shared.filter;


import org.dbtools.query.shared.QueryBuilder;

import javax.annotation.Nonnull;

public class LikeFilter extends CompareFilter {

    protected boolean ignoreCase;

    public static LikeFilter create(String field, Object value) {
        LikeFilter filterFormatter = new LikeFilter();
        filterFormatter.filter = LikeFilter.newInstance(field, value);
        return filterFormatter;
    }

    public static LikeFilter create(String field, Object value, boolean ignoreCase) {
        LikeFilter filterFormatter = new LikeFilter();
        filterFormatter.filter = LikeFilter.newInstance(field, value, ignoreCase);
        return filterFormatter;
    }

    private static LikeFilter newInstance(String field, Object value) {
        return new LikeFilter(field, value, true);
    }

    private static LikeFilter newInstance(String field, Object value, boolean ignoreCase) {
        return new LikeFilter(field, value, ignoreCase);
    }

    protected LikeFilter() {
        super();
    }

    private LikeFilter(String field, Object value, boolean ignoreCase) {
        super(field, null, value);
        this.ignoreCase = ignoreCase;
    }

    @Override
    public String build(@Nonnull QueryBuilder queryBuilder) {
        String clause;
        if (ignoreCase) {
            clause = queryBuilder.formatIgnoreCaseLikeClause(field, String.valueOf(queryBuilder.formatValue(value)));
        } else {
            clause = queryBuilder.formatLikeClause(field, String.valueOf(queryBuilder.formatValue(value)));
        }
        return clause;
    }

    public LikeFilter and(String field, Object value) {
        and(LikeFilter.create(field, value));
        return this;
    }

    public LikeFilter and(String field, Object value, boolean ignoreCase) {
        and(LikeFilter.create(field, value, ignoreCase));
        return this;
    }

    public LikeFilter or(String field, Object value) {
        or(LikeFilter.create(field, value));
        return this;
    }

    public LikeFilter or(String field, Object value, boolean ignoreCase) {
        or(LikeFilter.create(field, value, ignoreCase));
        return this;
    }

    @Override
    public LikeFilter clone() {
        LikeFilter clone = (LikeFilter) super.clone();
        clone.ignoreCase = this.ignoreCase;
        return clone;
    }
}

File: dbtools-query/src/commonMain/kotlin/org/dbtools/query/shared/filter/LikeFilter.kt
package org.dbtools.query.shared.filter

import org.dbtools.query.shared.QueryBuilder

class LikeFilter(
    field: String,
    value: Any?,
    var ignoreCase: Boolean = true,
) : CompareFilter(field, null, value) {

    override fun build(queryBuilder: QueryBuilder): String {
        val clause = if (ignoreCase) {
            queryBuilder.formatIgnoreCaseLikeClause(field.toString(), queryBuilder.formatValue(value).toString())
        } else {
            queryBuilder.formatLikeClause(field.toString(), queryBuilder.formatValue(value).toString())
        }
        return clause
    }

    override fun and(field: String, value: Any): LikeFilter {
        and(create(field, value))
        return this
    }

    fun and(field: String, value: Any, ignoreCase: Boolean): LikeFilter {
        and(create(field, value, ignoreCase))
        return this
    }

    override fun or(field: String, value: Any): LikeFilter {
        or(create(field, value))
        return this
    }

    fun or(field: String, value: Any, ignoreCase: Boolean): LikeFilter {
        or(create(field, value, ignoreCase))
        return this
    }

    companion object {
        fun create(field: String, value: Any): LikeFilter {
            val filterFormatter = LikeFilter(field, value)
            filterFormatter.filter = newInstance(field, value)
            return filterFormatter
        }

        fun create(field: String, value: Any, ignoreCase: Boolean): LikeFilter {
            val filterFormatter = LikeFilter(field, value, ignoreCase)
            filterFormatter.filter = newInstance(field, value, ignoreCase)
            return filterFormatter
        }

        private fun newInstance(field: String, value: Any): LikeFilter {
            return LikeFilter(field, value, true)
        }

        private fun newInstance(field: String, value: Any, ignoreCase: Boolean): LikeFilter {
            return LikeFilter(field, value, ignoreCase)
        }
    }
}
